### PR TITLE
Portable mapping devices

### PR DESCRIPTION
**Pull request by Arokha**
So the story goes that I wanted to make it so you could have some kinda advanced teleporter where you click on a map and it teleports you there. I dunno who'd get it but I wanted to make it, so I upgraded tgui to be able to tell you in the nanomap components where the user clicked, like the x/y turf coordinates. Then I realized nanomaps aren't generated for anything but a hard-coded list of maps now, and got frustrated since that'd mean it only works on those maps. I don't want to generate those for every map and admin-loadable one and whatnot and have all these hard-coded values (why did you make them hard-coded! GRR).

So I was like "Ok I'll port holomaps fully from /vg/ because they have those holochip things and maybe clicking on those would work", so I did that, but then was like "eh this is boring, I get a map, who cares", so I kept adding more and more and more until we got the mapping devices below.

Equipping the device and using it in-hand adds a draggable hud element:

https://user-images.githubusercontent.com/15028025/118751362-2f366600-b82f-11eb-83b4-0ba8a04b5208.mp4

https://user-images.githubusercontent.com/15028025/118751334-2180e080-b82f-11eb-9e8c-625dab06139c.mp4

Going to a map that isn't mapped (in game terms, not OOC terms. The holomap thing maps every z-level present at server start, but you can define z-levels that 'aren't allowed' to be mapped):

https://user-images.githubusercontent.com/15028025/118751502-66a51280-b82f-11eb-8cc1-7e310903bb75.mp4

It has a backup sonar mode thing you can use on those zlevels. Admin-loaded levels won't even have the map to display and will show nothing when using sonar (unless the admin pokes SSholomaps to generate the map for it).

https://user-images.githubusercontent.com/15028025/118751544-7e7c9680-b82f-11eb-9d4c-f0188c9a1e14.mp4

There's various flavors, like this ERT one that shows different helmets for different ERT jobs (based on worn helmet or rig). The different models can't show each other's markers, so you can have separate teams or whatever.

There's also a set of marker beacon items that show up on the map when active, one beacon type per mapper type. You could sneak one into a bag or onto a person to track them or whatever. I dunno. Make up your own spy shenanigans!

You can also map in permanent icons onto the map to mark like, arrivals and escape or whatever, using the holomap marker datums and landmark object. Unlike the beacons those are not tangible things ingame, and just make an icon show up on maps.

I'm definitely open to optimization suggestions. This 'finite state machine' is more like a 'finite state mosh pit'. I'm not even sure what's going on in there anymore.

They use a battery, by the way. The sonar mode drains it twice as fast. Only powered-on units are visible. A normal battery in normal mode lasts 32 minutes on a charge (that's the lowest drain I can manage... it's only 1 unit of charge!).

The beacons don't use any power and last forever.

### CODE REVIEW

File: code/modules/holomap/mapper.dm [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+
+		// Marker beacon items
+		for(var/hb in mapping_beacons)
+			var/obj/item/device/holomap_beacon/HB = hb
+			if(HB.mapper_filter != mapper_filter)
+				continue
+			
+			var/turf/TB = get_turf(HB)
+			if(TU.z != T_z)
+				continue
```
Comment (Atermonera) [inline code/modules/holomap/mapper.dm:350]: Do we need a nullcheck for the turf here too?

**Review by Atermonera — COMMENTED** [review]
Sorry for the delayss

File: code/_onclick/hud/screen_objects.dm [hunk 2]
```diff
@@ -655,3 +655,218 @@
...
+	var/obj/screen/mapper/bg/bg
+	
+	var/obj/screen/mapper/frame/frame
+	var/obj/screen/mapper/powbutton/powbutton
+	var/obj/screen/mapper/mapbutton/mapbutton
+
+	var/obj/item/device/mapping_unit/owner
+	var/obj/screen/mapper/extras_holder/extras_holder
+
+/obj/screen/movable/mapper_holder/New(newloc, newowner)
```
[the remark this answers is not in the document]
Comment (Arokha) [reply code/_onclick/hud/screen_objects.dm]: Don't think it matters. I think Initialize() was added to help avoid blocking mapload with New() calls and preventing people from joining. These aren't going to load at server start. I changed it, regardless.

File: code/_onclick/hud/screen_objects.dm [hunk 3]
```diff
@@ -655,3 +655,218 @@
...
+
+/obj/screen/mapper/mapbutton
+	icon = 'icons/effects/gpshud.dmi'
+	icon_state = "mapbutton"
+	plane = PLANE_HOLOMAP_FRAME
+	mouse_opacity = 1
+
+/obj/screen/mapper/mapbutton/Click()
+	if(!usr.checkClickCooldown())
+		return 1
```
[the remark this answers is not in the document]
Comment (Arokha) [reply code/_onclick/hud/screen_objects.dm]: Changed, was copypaste

File: code/modules/holomap/mapper.dm [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+		if(HC.mapper_filter != mapper_filter)
+			continue
+		var/mob_indicator = HOLOMAP_ERROR
+		var/turf/TU = get_turf(HC)
+		if(TU.z != T_z)
+			continue
+		
+		// Marker not on a turf
+		if(!TU)
+			continue
```
[the remark this answers is not in the document]
Comment (Arokha) [reply code/modules/holomap/mapper.dm]: True!

File: code/modules/holomap/mapper.dm (hunk 1 above)
Comment (Arokha) [reply code/modules/holomap/mapper.dm]: This whole section was indented wrong, it should have been TB. not TU., fixed

**Pull request merged**